File: shared/templates/bland_altman_template.R
# bland_altman_template.R — Bland-Altman 图模板
# ==============================================
# 适用于：方法比较研究、测量一致性评估（如新旧仪器、不同阅片者）
#
# 依赖: ggplot2, dplyr, blandr (可选)
# 安装: install.packages(c("ggplot2", "dplyr", "blandr"))
#
# 作者: MSRA Team
# 版本: 0.1.0

library(ggplot2)
library(dplyr)

# ============================================================================
# 1. 基本 Bland-Altman 图
# ============================================================================

#' 绘制 Bland-Altman 图
#'
#' @param method1 方法1的测量值
#' @param method2 方法2的测量值
#' @param title 图表标题
#' @param x_label X 轴标签
#' @param y_label Y 轴标签
#' @param conf_int 是否显示置信区间带
#' @param proportional_bias 是否检查比例偏差（可选）
#'
#' @return list(plot = ggplot, stats = list)
plot_bland_altman <- function(method1, method2,
                               title = "Bland-Altman Plot",
                               x_label = "Mean of Two Methods",
                               y_label = "Difference (Method1 - Method2)",
                               conf_int = TRUE,
                               proportional_bias = FALSE) {

  # 计算平均值和差值
  mean_vals <- (method1 + method2) / 2
  diff_vals <- method1 - method2

  n <- length(diff_vals)
  mean_diff <- mean(diff_vals, na.rm = TRUE)
  sd_diff <- sd(diff_vals, na.rm = TRUE)

  # 一致性界限 (LoA)
  loa_upper <- mean_diff + 1.96 * sd_diff
  loa_lower <- mean_diff - 1.96 * sd_diff

  # LoA 的置信区间
  se_loa <- sqrt(3 * sd_diff^2 / n)
  loa_upper_ci_upper <- loa_upper + 1.96 * se_loa
  loa_upper_ci_lower <- loa_upper - 1.96 * se_loa
  loa_lower_ci_upper <- loa_lower + 1.96 * se_loa
  loa_lower_ci_lower <- loa_lower - 1.96 * se_loa

  # 计算在 LoA 内的比例
  within_loa <- sum(diff_vals >= loa_lower & diff_vals <= loa_upper, na.rm = TRUE)
  pct_within <- within_loa / n * 100

  # 计算重复系数 (Repeatability Coefficient)
  rc <- 1.96 * sqrt(2) * sd_diff

  df <- data.frame(
    mean = mean_vals,
    diff = diff_vals
  )

  p <- ggplot(df, aes(x = mean, y = diff)) +
    # 零线
    geom_hline(yintercept = 0, linetype = "dotted", color = "grey50",
               linewidth = 0.5) +
    # 平均差异线
    geom_hline(yintercept = mean_diff, color = "steelblue", linewidth = 1) +
    # 一致性界限
    geom_hline(yintercept = loa_upper, linetype = "dashed", color = "#E74C3C",
               linewidth = 0.8) +
    geom_hline(yintercept = loa_lower, linetype = "dashed", color = "#E74C3C",
               linewidth = 0.8) +
    # 置信区间带
    ...  # (conf_int 区域，后续实现)

  # 散点
  p <- p + geom_point(alpha = 0.6, size = 2.5, color = "steelblue") +
    labs(
      title = title,
      subtitle = sprintf(
        "Mean diff = %.2f (SD = %.2f)\nUpper LoA = %.2f, Lower LoA = %.2f\n%.1f%% within LoA",
        mean_diff, sd_diff, loa_upper, loa_lower, pct_within
      ),
      x = x_label,
      y = y_label
    ) +
    theme_minimal(base_size = 11) +
    theme(
      plot.title = element_text(face = "bold", hjust = 0.5),
      plot.subtitle = element_text(hjust = 0.5, color = "grey40", size = 9)
    )

  # 可选：添加 LoA 置信区间带
  if (conf_int) {
    p <- p +
      annotate("rect",
               xmin = -Inf, xmax = Inf,
               ymin = loa_lower_ci_lower, ymax = loa_lower_ci_upper,
               fill = "#E74C3C", alpha = 0.05) +
      annotate("rect",
               xmin = -Inf, xmax = Inf,
               ymin = loa_upper_ci_lower, ymax = loa_upper_ci_upper,
               fill = "#E74C3C", alpha = 0.05)
  }

  # 可选：比例偏差检查
  if (proportional_bias) {
    lm_fit <- lm(diff ~ mean, data = df)
    p_val <- summary(lm_fit)$coefficients[2, 4]
    if (p_val < 0.05) {
      p <- p +
        geom_smooth(method = "lm", se = TRUE, color = "darkorange",
                    linewidth = 0.8, alpha = 0.2) +
        annotate("text", x = max(mean_vals, na.rm = TRUE) * 0.8,
                 y = max(diff_vals, na.rm = TRUE) * 0.9,
                 label = sprintf("Proportional bias p = %.3f", p_val),
                 size = 3.5, color = "darkorange", hjust = 1)
    }
  }

  stats <- list(
    n = n,
    mean_diff = mean_diff,
    sd_diff = sd_diff,
    loa_upper = loa_upper,
    loa_lower = loa_lower,
    pct_within_loa = pct_within,
    rc = rc,
    loa_upper_ci = c(loa_upper_ci_lower, loa_upper_ci_upper),
    loa_lower_ci = c(loa_lower_ci_lower, loa_lower_ci_upper)
  )

  return(list(plot = p, stats = stats))
}


# ============================================================================
# 2. 百分比差异 Bland-Altman 图（比例差异较大时用）
# ============================================================================

#' 绘制百分比 Bland-Altman 图
#'
#' 当差异与测量值成正比时使用（如测量范围跨越数量级）
#'
#' @param method1,method2 两种方法的测量值
#' @param title 标题
#' @return list(plot, stats)
plot_bland_altman_pct <- function(method1, method2,
                                   title = "Bland-Altman Plot (% Difference)") {

  mean_vals <- (method1 + method2) / 2
  pct_diff <- (method1 - method2) / mean_vals * 100

  n <- length(pct_diff)
  mean_pct <- mean(pct_diff, na.rm = TRUE)
  sd_pct <- sd(pct_diff, na.rm = TRUE)

  loa_upper <- mean_pct + 1.96 * sd_pct
  loa_lower <- mean_pct - 1.96 * sd_pct

  within_loa <- sum(pct_diff >= loa_lower & pct_diff <= loa_upper, na.rm = TRUE) / n * 100

  df <- data.frame(mean = mean_vals, pct_diff = pct_diff)

  p <- ggplot(df, aes(x = mean, y = pct_diff)) +
    geom_hline(yintercept = 0, linetype = "dotted", color = "grey50") +
    geom_hline(yintercept = mean_pct, color = "steelblue", linewidth = 1) +
    geom_hline(yintercept = loa_upper, linetype = "dashed", color = "#E74C3C") +
    geom_hline(yintercept = loa_lower, linetype = "dashed", color = "#E74C3C") +
    geom_point(alpha = 0.6, size = 2.5, color = "steelblue") +
    labs(
      title = title,
      subtitle = sprintf("Mean %% diff = %.1f%% (SD = %.1f%%)\nLoA: %.1f%% to %.1f%%",
                         mean_pct, sd_pct, loa_lower, loa_upper),
      x = "Mean of Two Methods",
      y = "% Difference (Method1 - Method2) / Mean × 100"
    ) +
    theme_minimal(base_size = 11) +
    theme(
      plot.title = element_text(face = "bold", hjust = 0.5),
      plot.subtitle = element_text(hjust = 0.5, color = "grey40", size = 9)
    )

  return(list(
    plot = p,
    stats = list(
      mean_pct_diff = mean_pct,
      sd_pct_diff = sd_pct,
      loa_upper = loa_upper,
      loa_lower = loa_lower,
      pct_within = within_loa
    )
  ))
}


# ============================================================================
# 3. 分组 Bland-Altman 图
# ============================================================================

#' 分组 Bland-Altman 图（如不同阅片者、不同批次）
#'
#' @param method1,method2 测量值
#' @param group 分组变量
#' @param title 标题
#' @return ggplot 对象
plot_bland_altman_grouped <- function(method1, method2, group,
                                       title = "Bland-Altman Plot by Group") {

  mean_vals <- (method1 + method2) / 2
  diff_vals <- method1 - method2

  df <- data.frame(
    mean = mean_vals,
    diff = diff_vals,
    group = factor(group)
  )

  # 计算每组统计量
  group_stats <- df %>%
    group_by(group) %>%
    summarise(
      n = n(),
      mean_diff = mean(diff, na.rm = TRUE),
      sd_diff = sd(diff, na.rm = TRUE),
      loa_upper = mean_diff + 1.96 * sd_diff,
      loa_lower = mean_diff - 1.96 * sd_diff,
      .groups = "drop"
    )

  p <- ggplot(df, aes(x = mean, y = diff, color = group)) +
    geom_point(alpha = 0.6, size = 2) +
    geom_hline(yintercept = 0, linetype = "dotted", color = "grey50") +
    # 各组的平均差异线
    geom_hline(data = group_stats,
               aes(yintercept = mean_diff, color = group),
               linewidth = 1, alpha = 0.8) +
    # 一致性界限（使用虚线，各组的稍微偏移以避免重叠）
    geom_hline(data = group_stats,
               aes(yintercept = loa_upper, color = group),
               linetype = "dashed", linewidth = 0.6, alpha = 0.6) +
    geom_hline(data = group_stats,
               aes(yintercept = loa_lower, color = group),
               linetype = "dashed", linewidth = 0.6, alpha = 0.6) +
    labs(
      title = title,
      x = "Mean of Two Methods",
      y = "Difference"
    ) +
    theme_minimal(base_size = 11) +
    theme(
      plot.title = element_text(face = "bold", hjust = 0.5),
      legend.position = "bottom"
    )

  attr(p, "group_stats") <- group_stats
  return(p)
}


# ============================================================================
# 示例
# ============================================================================
if (FALSE) {
  set.seed(42)
  n <- 100

  # 模拟数据
  true_val <- runif(n, 10, 100)
  method1 <- true_val + rnorm(n, 0, 3)
  method2 <- true_val + rnorm(n, 2, 4)  # 有系统偏差

  # 基本 Bland-Altman
  res <- plot_bland_altman(method1, method2,
                           title = "Device A vs Device B")
  print(res$plot)
  print(res$stats)
  ggsave("bland_altman.png", res$plot, width = 8, height = 7, dpi = 300)

  # 百分比 Bland-Altman（当差异与值成比例时）
  method1_prop <- true_val * rnorm(n, 1, 0.05) + 2
  method2_prop <- true_val * rnorm(n, 0.95, 0.08) + 3
  res2 <- plot_bland_altman_pct(method1_prop, method2_prop)
  ggsave("bland_altman_pct.png", res2$plot, width = 8, height = 7, dpi = 300)

  cat("✅ Bland-Altman 模板示例完成\n")
}

cat("✅ bland_altman_template.R 已加载\n")
cat("可用函数:\n")
cat("  plot_bland_altman()          - 基本 BA 图\n")
cat("  plot_bland_altman_pct()      - 百分比差异 BA 图\n")
cat("  plot_bland_altman_grouped()  - 分组 BA 图\n")

File: shared/templates/bland_altman_template.py
"""
bland_altman_template.py — Bland-Altman 图模板（Python）
=======================================================
适用于：方法比较、测量一致性评估

依赖: matplotlib, numpy, pandas, scipy
安装: pip install matplotlib numpy pandas scipy

作者: MSRA Team
版本: 0.1.0
"""

from typing import Dict, Optional, Tuple

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy import stats
import logging

logger = logging.getLogger(__name__)


def plot_bland_altman(
    method1: np.ndarray,
    method2: np.ndarray,
    title: str = "Bland-Altman Plot",
    x_label: str = "Mean of Two Methods",
    y_label: str = "Difference (Method1 - Method2)",
    conf_int: bool = True,
    proportional_bias: bool = False,
    figsize: Tuple[float, float] = (8, 6),
) -> Dict:
    """Bland-Altman 图（绝对差异）

    Parameters
    ----------
    method1, method2 : np.ndarray
        两种方法的测量值
    title, x_label, y_label : str
        标题和轴标签
    conf_int : bool
        是否显示 LoA 置信区间带
    proportional_bias : bool
        是否检查比例偏差
    figsize : Tuple
        图尺寸

    Returns
    -------
    Dict
        包含 fig, ax, stats
    """
    mean_vals = (method1 + method2) / 2
    diff_vals = method1 - method2

    n = len(diff_vals)
    mean_diff = np.nanmean(diff_vals)
    sd_diff = np.nanstd(diff_vals, ddof=1)

    loa_upper = mean_diff + 1.96 * sd_diff
    loa_lower = mean_diff - 1.96 * sd_diff

    # LoA 置信区间
    se_loa = np.sqrt(3 * sd_diff**2 / n)
    loa_upper_ci = (loa_upper - 1.96 * se_loa, loa_upper + 1.96 * se_loa)
    loa_lower_ci = (loa_lower - 1.96 * se_loa, loa_lower + 1.96 * se_loa)

    within_loa = np.sum((diff_vals >= loa_lower) & (diff_vals <= loa_upper))
    pct_within = within_loa / n * 100

    # 重复系数
    rc = 1.96 * np.sqrt(2) * sd_diff

    fig, ax = plt.subplots(figsize=figsize)

    # 散点
    ax.scatter(mean_vals, diff_vals, alpha=0.6, s=30, color="steelblue",
               edgecolors="white", linewidth=0.5)

    # 平均差异线
    ax.axhline(y=mean_diff, color="steelblue", linewidth=1.5, label=f"Mean diff = {mean_diff:.2f}")

    # 一致性界限
    ax.axhline(y=loa_upper, linestyle="--", color="#E74C3C", linewidth=1,
               label=f"Upper LoA = {loa_upper:.2f}")
    ax.axhline(y=loa_lower, linestyle="--", color="#E74C3C", linewidth=1,
               label=f"Lower LoA = {loa_lower:.2f}")

    # 零线
    ax.axhline(y=0, linestyle=":", color="grey", linewidth=0.8, alpha=0.5)

    # LoA 置信区间带
    if conf_int:
        ax.axhspan(loa_lower_ci[0], loa_lower_ci[1],
                   alpha=0.05, color="#E74C3C")
        ax.axhspan(loa_upper_ci[0], loa_upper_ci[1],
                   alpha=0.05, color="#E74C3C")

    # 比例偏差检查
    if proportional_bias:
        slope, intercept, r_val, p_val, std_err = stats.linregress(mean_vals, diff_vals)
        if p_val < 0.05:
            x_line = np.array([mean_vals.min(), mean_vals.max()])
            y_line = intercept + slope * x_line
            ax.plot(x_line, y_line, color="darkorange", linewidth=1, linestyle="-",
                    alpha=0.7, label=f"Proportional bias (p={p_val:.3f})")

    ax.set_xlabel(x_label, fontsize=11)
    ax.set_ylabel(y_label, fontsize=11)
    ax.set_title(title, fontsize=13, fontweight="bold")

    subtitle_text = (
        f"Mean diff = {mean_diff:.2f} (SD = {sd_diff:.2f})\n"
        f"Upper LoA = {loa_upper:.2f}, Lower LoA = {loa_lower:.2f}\n"
        f"{pct_within:.1f}% within LoA | RC = {rc:.2f}"
    )
    ax.text(0.98, 0.02, subtitle_text, transform=ax.transAxes,
            fontsize=8, va="bottom", ha="right",
            bbox=dict(boxstyle="round", facecolor="white", alpha=0.8))

    ax.legend(fontsize=8, loc="upper left")
    ax.grid(alpha=0.3, linestyle=":")

    plt.tight_layout()

    return {
        "fig": fig,
        "ax": ax,
        "stats": {
            "n": n,
            "mean_diff": mean_diff,
            "sd_diff": sd_diff,
            "loa_upper": loa_upper,
            "loa_lower": loa_lower,
            "pct_within_loa": pct_within,
            "rc": rc,
        },
    }


def plot_bland_altman_pct(
    method1: np.ndarray,
    method2: np.ndarray,
    title: str = "Bland-Altman Plot (% Difference)",
    figsize: Tuple[float, float] = (8, 6),
) -> Dict:
    """百分比差异 Bland-Altman 图"""
    mean_vals = (method1 + method2) / 2
    pct_diff = (method1 - method2) / mean_vals * 100

    n = len(pct_diff)
    mean_pct = np.nanmean(pct_diff)
    sd_pct = np.nanstd(pct_diff, ddof=1)

    loa_upper = mean_pct + 1.96 * sd_pct
    loa_lower = mean_pct - 1.96 * sd_pct

    within = np.sum((pct_diff >= loa_lower) & (pct_diff <= loa_upper)) / n * 100

    fig, ax = plt.subplots(figsize=figsize)
    ax.scatter(mean_vals, pct_diff, alpha=0.6, s=30, color="steelblue")
    ax.axhline(y=mean_pct, color="steelblue", linewidth=1.5)
    ax.axhline(y=loa_upper, linestyle="--", color="#E74C3C", linewidth=1)
    ax.axhline(y=loa_lower, linestyle="--", color="#E74C3C", linewidth=1)
    ax.axhline(y=0, linestyle=":", color="grey", linewidth=0.8)
    ax.set_xlabel("Mean of Two Methods", fontsize=11)
    ax.set_ylabel("% Difference", fontsize=11)
    ax.set_title(title, fontsize=13, fontweight="bold")

    stats_text = (
        f"Mean % diff = {mean_pct:.1f}% (SD = {sd_pct:.1f}%)\n"
        f"LoA: {loa_lower:.1f}% to {loa_upper:.1f}%\n"
        f"{within:.1f}% within LoA"
    )
    ax.text(0.98, 0.02, stats_text, transform=ax.transAxes,
            fontsize=9, va="bottom", ha="right",
            bbox=dict(boxstyle="round", facecolor="white", alpha=0.8))
    ax.grid(alpha=0.3, linestyle=":")
    plt.tight_layout()

    return {
        "fig": fig,
        "stats": {
            "mean_pct_diff": mean_pct,
            "sd_pct_diff": sd_pct,
            "loa_upper": loa_upper,
            "loa_lower": loa_lower,
            "pct_within_loa": within,
        },
    }


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format='%(levelname)s: %(message)s')
    np.random.seed(42)
    n = 100
    true_val = np.random.uniform(10, 100, n)
    m1 = true_val + np.random.normal(0, 3, n)
    m2 = true_val + np.random.normal(2, 4, n)  # bias

    res = plot_bland_altman(m1, m2, title="Device A vs Device B")
    res["fig"].savefig("bland_altman_demo.png", dpi=150, bbox_inches="tight")
    plt.close()
    logger.info("✅ bland_altman_demo.png 已保存")
    logger.info(f"Stats: mean_diff={res['stats']['mean_diff']:.2f}")
